### PR TITLE
Dune build rules for doc_grammar and fullGrammar.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** infrastructure.

cc @jfehrle 

Notes:
- I cannot go further and have a rule update, e.g. `orderedGrammar` or the rst files because Dune won't accept a file being both a dependency and a target of a rule. We need to change the model this tool follows to be compatible with Dune. For the rst files, we could generate files containing grammar chunks only and include them (using the appropriate Sphinx directive) in the main rst files.
- We cannot have multiple rules generating the same target. It would be better if the tool was able to use `fullGrammar` as input for the next steps.
- I didn't find any better way of providing the dependencies than listing them all explicitly.

### CODE REVIEW

File: doc/tools/docgram/dune [hunk 1]
```diff
@@ -0,0 +1,10 @@
...
+ (name doc_grammar)
+ (libraries coq.clib coqpp))
+
+(rule
+ (targets fullGrammar)
+ (deps ../../../parsing/g_constr.mlg ../../../parsing/g_prim.mlg ../../../toplevel/g_toplevel.mlg ../../../vernac/g_proofs.mlg ../../../vernac/g_vernac.mlg ../../../plugins/btauto/g_btauto.mlg ../../../plugins/cc/g_congruence.mlg ../../../plugins/derive/g_derive.mlg ../../../plugins/extraction/g_extraction.mlg ../../../plugins/firstorder/g_ground.mlg ../../../plugins/funind/g_indfun.mlg ../../../plugins/ltac/coretactics.mlg ../../../plugins/ltac/extraargs.mlg ../../../plugins/ltac/extratactics.mlg ../../../plugins/ltac/g_auto.mlg ../../../plugins/ltac/g_class.mlg ../../../plugins/ltac/g_eqdecide.mlg ../../../plugins/ltac/g_ltac.mlg ../../../plugins/ltac/g_obligations.mlg ../../../plugins/ltac/g_rewrite.mlg ../../../plugins/ltac/g_tactic.mlg ../../../plugins/ltac/profile_ltac_tactics.mlg ../../../plugins/micromega/g_micromega.mlg ../../../plugins/micromega/g_zify.mlg ../../../plugins/nsatz/g_nsatz.mlg ../../../plugins/omega/g_omega.mlg ../../../plugins/rtauto/g_rtauto.mlg ../../../plugins/setoid_ring/g_newring.mlg ../../../plugins/syntax/g_numeral.mlg ../../../plugins/syntax/g_string.mlg)
```
Comment (ejgallego) [inline doc/tools/docgram/dune]: Maybe you want to use a `glob` here?

File: doc/tools/docgram/dune [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+ (action
+  (chdir ../../.. (run doc/tools/docgram/doc_grammar.exe -short -no-warn %{deps})))
```
Comment (ejgallego) [inline doc/tools/docgram/dune]: Umm, maybe you want to use `${project_root}` ?

**Comment by ejgallego** [comment]
> * I cannot go further and have a rule update, e.g. `orderedGrammar` or the rst files because Dune won't accept a file being both a dependency and a target of a rule.

Can't you just generate a `.new` file and use the diff action at the end?

> * We cannot have multiple rules generating the same target. It would be better if the tool was able to use `fullGrammar` as input for the next steps.

That's indeed problematic IMO, you can have a rule calling some program / using some dependency conditionally tho as long as you generate the same set of targets [targets are kinda static in the selective functor dune model]

> * I didn't find any better way of providing the dependencies than listing them all explicitly.

As I suggested you could try using a glob, let me know if it doesn't work.

File: doc/tools/docgram/dune (hunk 1 above)
Comment (Zimmi48) [reply doc/tools/docgram/dune]: That shortens things a bit, thanks! But the glob language is too limited to make things as short as in `Makefile.doc`.

File: doc/tools/docgram/dune (hunk 2 above)
Comment (Zimmi48) [reply doc/tools/docgram/dune]: Doesn't work :confused:

**Comment by Zimmi48** [comment]
> Can't you just generate a `.new` file and use the diff action at the end?

That could work, thanks for the suggestion. But this still needs some tweak to `doc_grammar.ml`.

File: doc/tools/docgram/dune (hunk 2 above)
Comment (ejgallego) [reply doc/tools/docgram/dune]: What error do you get? Also, you may want to add `doc_grammar` exe to the local binaries `(env )` and thus avoid having to deal with the path.

File: doc/tools/docgram/dune (hunk 1 above)
Comment (ejgallego) [reply doc/tools/docgram/dune]: I was under the impression that `**.mlg` did work in Dune.

**Comment by jfehrle** [comment]
(I almost missed this, I get so much mail for \@doc_maintainers and all I saw of the title was "Dune build rules".)

> We cannot have multiple rules generating the same target. It would be better if the tool was able to use fullGrammar as input for the next steps.

That's already there, that's what "make doc_gram" does.  It doesn't touch orderedGrammar.

> I cannot go further and have a rule update, e.g. `orderedGrammar` or the rst files because Dune won't accept a file being both a dependency and a target of a rule. 

What if doc_grammar creates an empty flag file `updated_orderedGrammar` whenever it updates `orderedGrammar` and we just don't tell Dune that orderedGrammar is updated?  Then the rst files can depend on the flag file.  We tell Dune that the flag file depends on `orderedGrammar`. 

Similarly, when doc_grammar updates the rsts, it already creates the flag file `updated_rsts` that Sphinx could depend on, but Dune would not know that doc_grammar updates the rsts.  We tell Dune that `rsts_updated` depends on orderedGrammar and all the rsts.

Is that plausible or fatally flawed?

> For the rst files, we could generate files containing grammar chunks only and include them (using the appropriate Sphinx directive) in the main rst files.

I think it would be quite cumbersome to have a separate file for each `insertprodn`.  The extra indirection will make it much harder to see the productions that appear, both while editing and in reviews.  I have 70 insertprodns in my converted grammar.  The Dune dependencies for the relevant rsts would need to be updated every time you add another insertprodn.  Is there an automated process for doings?

In addition, we want doc_grammar to update syntax for all the commands and tactics.  That would be about six hundred more files.

Is it really necessary to change the orderedGrammar and rst behavior right now?  I'd prefer not to make big changes while we're in the middle of getting the syntax in the doc.  Converting later, if we choose to do that, should be readily automatable with some throwaway code.

File: doc/tools/docgram/dune (hunk 1 above)
Comment (Zimmi48) [reply doc/tools/docgram/dune]: It exists, but it doesn't do what you think. See https://dune.readthedocs.io/en/stable/concepts.html#glob, and in particular:

>Currently dune only supports globbing files in a single directory.

File: doc/tools/docgram/dune (hunk 2 above)
Comment (Zimmi48) [reply doc/tools/docgram/dune]: Actually, it does work, but with the `%{project_root}` syntax. I tested your suggestion too quickly without realizing the typo.

**Comment by Zimmi48** [comment]
@jfehrle 

> Is it really necessary to change the orderedGrammar and rst behavior right now?

Absolutely not. I just raised the question because as you asked me to get my hands dirty with docgram, I finally did what I had promised and looked into building it with Dune.

Also, note that this PR is good to go as is: it stops at building the tool and the `fullGrammar` file, but that's still useful enough.

> > We cannot have multiple rules generating the same target. It would be better if the tool was able to use fullGrammar as input for the next steps.
> 
> That's already there, that's what "make doc_gram" does. It doesn't touch orderedGrammar.

You misunderstood the suggestion. I know of course that there is a way to only build `fullGrammar` since this is exactly what this PR does. What I don't like and that Dune doesn't like either is the fact that the next step (generating `editedGrammar`) cannot be done using the output generated by the previous step (`fullGrammar`) as an input. Or if it is supported, it isn't documented and it isn't used in `Makefile.doc`.

There is a way around that, of course, which is not to tell Dune that the rule to build `editedGrammar`  does also generate `fullGrammar`. But that's not pretty because that means that if a user asks to build `fullGrammar`, then `editedGrammar`, then `fullGrammar`, the last step won't be a no-op.

**Comment by Zimmi48** [comment]
@ejgallego 

> > * I cannot go further and have a rule update, e.g. `orderedGrammar` or the rst files because Dune won't accept a file being both a dependency and a target of a rule.
> 
> Can't you just generate a `.new` file and use the diff action at the end?

In fact, this doesn't work because Dune understands that `orderedGrammar` is a target even if it is not explicitly listed as such:

The following which emulates what `doc_grammar` would need to do to follow this approach:

```
(rule
 (alias update-orderedGrammar)
 (deps
  (:input
   ; Main grammar
   (glob_files %{project_root}/parsing/*.mlg)
   (glob_files %{project_root}/toplevel/*.mlg)
   (glob_files %{project_root}/vernac/*.mlg)
   ; All plugins except SSReflect for now (mimicking what is done in Makefile.doc)
   (glob_files %{project_root}/plugins/btauto/*.mlg)
   (glob_files %{project_root}/plugins/cc/*.mlg)
   (glob_files %{project_root}/plugins/derive/*.mlg)
   (glob_files %{project_root}/plugins/extraction/*.mlg)
   (glob_files %{project_root}/plugins/firstorder/*.mlg)
   (glob_files %{project_root}/plugins/funind/*.mlg)
   (glob_files %{project_root}/plugins/ltac/*.mlg)
   (glob_files %{project_root}/plugins/micromega/*.mlg)
   (glob_files %{project_root}/plugins/nsatz/*.mlg)
   (glob_files %{project_root}/plugins/omega/*.mlg)
   (glob_files %{project_root}/plugins/rtauto/*.mlg)
   (glob_files %{project_root}/plugins/setoid_ring/*.mlg)
   (glob_files %{project_root}/plugins/syntax/*.mlg))
  orderedGrammar)
 (action
  (progn
   (copy orderedGrammar orderedGrammar.old)
   (chdir %{project_root}
    (progn
     (run doc_grammar %{input})
     (chdir doc/tools/docgram
      (progn
       (copy orderedGrammar orderedGrammar.new)
       (copy orderedGrammar.old orderedGrammar)
       (diff orderedGrammar orderedGrammar.new))))))))
```

gives the following error:

```
Error: Multiple rules generated for
_build/default/doc/tools/docgram/orderedGrammar:
- file present in source tree
- doc/tools/docgram/dune:7
Hint: rm -f doc/tools/docgram/orderedGrammar
```

and the same thing with `(mode promote)`:

```
(rule
 (alias update-orderedGrammar)
 (deps
  (:input
   ; Main grammar
   (glob_files %{project_root}/parsing/*.mlg)
   (glob_files %{project_root}/toplevel/*.mlg)
   (glob_files %{project_root}/vernac/*.mlg)
   ; All plugins except SSReflect for now (mimicking what is done in Makefile.doc)
   (glob_files %{project_root}/plugins/btauto/*.mlg)
   (glob_files %{project_root}/plugins/cc/*.mlg)
   (glob_files %{project_root}/plugins/derive/*.mlg)
   (glob_files %{project_root}/plugins/extraction/*.mlg)
   (glob_files %{project_root}/plugins/firstorder/*.mlg)
   (glob_files %{project_root}/plugins/funind/*.mlg)
   (glob_files %{project_root}/plugins/ltac/*.mlg)
   (glob_files %{project_root}/plugins/micromega/*.mlg)
   (glob_files %{project_root}/plugins/nsatz/*.mlg)
   (glob_files %{project_root}/plugins/omega/*.mlg)
   (glob_files %{project_root}/plugins/rtauto/*.mlg)
   (glob_files %{project_root}/plugins/setoid_ring/*.mlg)
   (glob_files %{project_root}/plugins/syntax/*.mlg))
  orderedGrammar)
 (action
  (progn
   (copy orderedGrammar orderedGrammar.old)
   (chdir %{project_root}
    (progn
     (run doc_grammar %{input})
     (chdir doc/tools/docgram
      (progn
       (copy orderedGrammar orderedGrammar.new)
       (copy orderedGrammar.old orderedGrammar)
       (diff orderedGrammar orderedGrammar.new)))))))
 (mode promote))
```

gives this one:

```
Error: Dependency cycle between the following files:
   _build/default/doc/tools/docgram/orderedGrammar
```

And one last attempt was:

```
(rule
 (alias update-orderedGrammar)
 (deps
  (:input
   ; Main grammar
   (glob_files %{project_root}/parsing/*.mlg)
   (glob_files %{project_root}/toplevel/*.mlg)
   (glob_files %{project_root}/vernac/*.mlg)
   ; All plugins except SSReflect for now (mimicking what is done in Makefile.doc)
   (glob_files %{project_root}/plugins/btauto/*.mlg)
   (glob_files %{project_root}/plugins/cc/*.mlg)
   (glob_files %{project_root}/plugins/derive/*.mlg)
   (glob_files %{project_root}/plugins/extraction/*.mlg)
   (glob_files %{project_root}/plugins/firstorder/*.mlg)
   (glob_files %{project_root}/plugins/funind/*.mlg)
   (glob_files %{project_root}/plugins/ltac/*.mlg)
   (glob_files %{project_root}/plugins/micromega/*.mlg)
   (glob_files %{project_root}/plugins/nsatz/*.mlg)
   (glob_files %{project_root}/plugins/omega/*.mlg)
   (glob_files %{project_root}/plugins/rtauto/*.mlg)
   (glob_files %{project_root}/plugins/setoid_ring/*.mlg)
   (glob_files %{project_root}/plugins/syntax/*.mlg))
  orderedGrammar)
 (action
  (progn
   (copy orderedGrammar orderedGrammar.old)
   (chdir %{project_root}
    (progn
     (run doc_grammar %{input})
     (chdir doc/tools/docgram
      (diff orderedGrammar.old orderedGrammar)))))))
```

This rule works as long as there is no change to the parser. If there is one and that the diff fails, it is printed as:

```
Done: 334/336 (jobs: 1)File "doc/tools/docgram/orderedGrammar.old", line 1, characters 0-0:
         git (internal) (exit 1)
(cd _build/default && /nix/store/4d34zc1w0s7dd8z9mjd69962gj13zkwp-git-2.25.1/bin/git diff --no-index --color=always -u doc/tools/docgram/orderedGrammar.old doc/tools/docgram/orderedGrammar)
diff --git a/doc/tools/docgram/orderedGrammar.old b/doc/tools/docgram/orderedGrammar
index f26a174722..cddcf1227a 100644
--- a/doc/tools/docgram/orderedGrammar.old
+++ b/doc/tools/docgram/orderedGrammar
```

and doesn't propose to promote the file (which isn't surprising in fact, how would it know how to promote it).

File: doc/tools/docgram/dune (hunk 1 above)
Comment (Zimmi48) [reply doc/tools/docgram/dune]: See also ocaml/dune#1866

**Comment by Zimmi48** [comment]
@jfehrle 

> What if doc_grammar creates an empty flag file `updated_orderedGrammar` whenever it updates `orderedGrammar` and we just don't tell Dune that `orderedGrammar` is updated?

I think you are forgetting about one of the most important feature of Dune: hygienic builds. `orderedGrammar` is *not* updated, only a copy is. Unless we ask Dune to "promote" the file, but in this case, it understands that it is both a target and a dependency and it doesn't accept this.

> I think it would be quite cumbersome to have a separate file for each `insertprodn`. The extra indirection will make it much harder to see the productions that appear, both while editing and in reviews. I have 70 insertprodns in my converted grammar. The Dune dependencies for the relevant rsts would need to be updated every time you add another insertprodn. Is there an automated process for doings?
> 
> In addition, we want doc_grammar to update syntax for all the commands and tactics. That would be about six hundred more files.

You are right. Then, there is no hope to finish the Dune integration of this tool. We'll have to keep a small Makefile to be able to update `orderedGrammar` and the `rst` when needed. To be able to use Dune for this, we would need to get a feature that would allow to regenerate an existing file, using this file as an input (i.e. a "bootstrap promote" mode).

**Comment by jfehrle** [comment]
> Absolutely not. I just raised the question because as you asked me to get my hands dirty with docgram, I finally did what I had promised and looked into building it with Dune.

Much appreciated, thanks!

> You misunderstood the suggestion. I know of course that there is a way to only build fullGrammar since this is exactly what this PR does. What I don't like and that Dune doesn't like either is the fact that the next step (generating editedGrammar) cannot be done using the output generated by the previous step (fullGrammar) as an input. Or if it is supported, it isn't documented and it isn't used in Makefile.doc.

editedGrammar is unimportant  I haven't looked at that file for a long while.  orderedGrammar matters.  It would be simple to add a command line option that reads fullGrammar instead of generating it.  But the point is moot because we need to keep the makefile for other reasons.

> I think you are forgetting about one of the most important feature of Dune

I don't know a lot about Dune.  I'm not using it yet.

> Then, there is no hope to finish the Dune integration of this tool. We'll have to keep a small Makefile ...

In our earlier discussions, we planned to have CI use the -verify option ("make doc_gram_verify"), which doesn't update any files at all, but gives an error if they don't match freshly-generated versions of those files.  The updating of orderedGrammar and the rsts is intended for developer use on their local system.  Perhaps we could take advantage of this at some point.

File: doc/tools/docgram/dune (hunk 1 above)
Comment (ejgallego) [reply doc/tools/docgram/dune]: Oh indeed.

File: doc/tools/docgram/dune (hunk 2 above)
Comment (ejgallego) [reply doc/tools/docgram/dune]: Oh, sorry for the typo.

**Comment by ejgallego** [comment]
Umm, it is hard to follow what is going here, the code of `doc_grammar` is also not very easy to follow.

IMO what you are trying to do should not be hard, provided you follow some rules regarding promoted objects, I suggest we discuss more in person tho.

[@Zimmi48 , this can be merged it seems so let me know when you want it in]

**Comment by jfehrle** [comment]
@ejgallego > the code of `doc_grammar` is also not very easy to follow

Then please ask questions.  Like a lot of code, it's easier when you understand the patterns and harder when you try to infer them.

**Comment by Zimmi48** [comment]
> [@Zimmi48 , this can be merged it seems so let me know when you want it in]

Yes, please merge. I don't think we will do any more progress for now.

**Review by ejgallego — APPROVED** [review]
